synapseclient/api/entity_factory.py

```diff
@@ -0,0 +1,391 @@
+"""Factory type functions to create and retrieve entities from Synapse"""
+
+import os
+from typing import TYPE_CHECKING, Any, Dict, Optional, Union
+
+from opentelemetry import trace
+
+from synapseclient.api.entity_bundle_services_v2 import (
+    get_entity_id_bundle2,
+    get_entity_id_version_bundle2,
+)
+from synapseclient.api.entity_services import get_entities_by_md5, get_entity_path
+from synapseclient.core import utils
+from synapseclient.core.constants import concrete_types
+from synapseclient.core.download import download_file_entity_model
+from synapseclient.core.exceptions import (
+    SynapseFileNotFoundError,
+    SynapseUnmetAccessRestrictions,
+)
+
+if TYPE_CHECKING:
+    from synapseclient import Synapse
+    from synapseclient.models import File, Folder, Project
+
+
+async def get_from_entity_factory(
+    synapse_id_or_path: str,
+    version: int = None,
+    if_collision: str = "keep.both",
+    limit_search: str = None,
+    md5: str = None,
+    download_file: bool = True,
+    download_location: str = None,
+    follow_link: bool = False,
+    entity_to_update: Union["Project", "File", "Folder"] = None,
+    *,
```

Comment:
I like this factory class!  It's a lot more readable than the old code was as it has a clear series of steps that it does.

Just curious, what was the rationale behind the `*`?

Reply:
@Sage-Bionetworks/dpe This would be a great spot to get some feedback from the team.

What the asterisk in the parameters means:

> You use a bare asterisk to define a boundary between arguments that you can pass by either position or keyword from those that you must pass by keyword.
![image](https://github.com/Sage-Bionetworks/synapsePythonClient/assets/17128019/f8fa98f6-018e-4f9f-acd9-fe19b18bc89c)


Here my my thought process:

1. By requiring that `synapse_client` is always a KW only argument it keeps a consistent experience across every function.
2. Passing `synapse_client` into a function is not meant to be the norm, but rather the exception to allow for you to use multiple logged in synapse client sessions during a single script execution run. Providing this functionality is to allow for a use case where a single running server instance might be providing the interface for multiple user accounts.
3. In the future when we want to add more arguments to a function then we will never have a case where `synapse_client` ends up being in the "middle" of several other positional arguments.

What is everyone thought on using `*` across the board for this purpose? Is this something you support, think would be confusing, or is a bad decision?

Reply:
@BryanFauble I think the reasoning is sound and I'm for turning it into a kw-only arg, so long as the documentation examples using the `synapse_client` keyword are updated to explain this. I imagine it would be a bit of a clumsy transition for users who are not as experienced with keyword-only args and already have pre-existing scripts where they use `synapse_client` in the old way.